test/utils.py

```diff
@@ -79,3 +81,24 @@ def wait_file(dir_name, file_pattern, exclude_list, timeout_sec=1):
         iter_count = iter_count + 1
 
     return ""
+
+
+def kill_instance(instance_file_name):
+    # Kill tarantool instance by file config name.
+    pid_instance = subprocess.run(["pgrep", instance_file_name], capture_output=True)
+    if len(pid_instance.stdout) != 0:
+        os.kill(int(pid_instance.stdout), signal.SIGKILL)
+
+
+def find_free_port():
+    # Return free port in range [49152, 65535], else return None.
+    for port in range(49152, 65536):
```

Comment:
It is from the local port range:

```sh
$ cat /proc/sys/net/ipv4/ip_local_port_range
32768	60999
```

From the kernel [documentation](https://www.kernel.org/doc/Documentation/networking/ip-sysctl.rst):

```
ip_local_port_range - 2 INTEGERS
	Defines the local port range that is used by TCP and UDP to
	choose the local port. The first number is the first, the
	second the last local port number.
	If possible, it is better these numbers have different parity
	(one even and one odd value).
	Must be greater than or equal to ip_unprivileged_port_start.
	The default values are 32768 and 60999 respectively.
```

In other words, when a client calls `connect()` (without prior explicit `bind()`, it is usual for a client to do just `connect()`), it'll choose a port from this range. The following sequence of events is possible so:

* You checked a port, it is free. You write it to a file.
* A client connects using TCP to some server.
* You attempt to listen on this port, but it is already stolen.

Unlikely you'll see this situation in a single thread testing, but if we'll use `xdist` in a future, it is possible. We observed the similar problem in test-run (https://github.com/tarantool/test-run/issues/141), where the port assignment mechanism is quite similar (except that we pass the address to tarantool using an environment variable).

I have a proposal how to stick with bind-to-zero-and-listen approach even on 1.10, will write below.

Reply:
Thanks for the feedback! Now I have made an implementation using `box.cfg({listen = '127.0.0.1:0'})`. The port is obtained by `box.info.listen` or (if `box.info.listen = nil`) with using snippet below. Pytest gets binded port from file inside temp dir during testing.